pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go

```diff
@@ -31,265 +30,183 @@ import (
-							needUpdate = true
-						}
-					}
-				}
-				if needUpdate {
-					p.agentUpdateChan <- agentUpdateEvent
-				}
 			}
-		}
-	}()
-}
 
-func (p *AgentPoller) Stop() {
-	if p.agentDone != nil {
-		close(p.agentDone)
-		p.agentDone = nil
+		}
 	}
-}
-
-func (p *AgentPoller) GetInterfaceStatuses(cmdResult string) []api.InterfaceStatus {
-	parsedResult := parseAgentReplyToJson(cmdResult)
-	interfaceStatuses := calculateInterfaceStatusesFromAgentJson(parsedResult)
-	return p.mergeAgentStatusesWithDomainData(interfaceStatuses)
-}
 
-func (p *AgentPoller) GetGuestOsInfo(cmdResult string) api.GuestOSInfo {
-	osInfo := parseAgentGuestInfoReplyToJson(cmdResult)
-	guestInfo := api.GuestOSInfo{}
-	guestInfo.Name = osInfo.Name
-	guestInfo.KernelRelease = osInfo.KernelRelease
-	guestInfo.Version = osInfo.Version
-	guestInfo.PrettyName = osInfo.PrettyName
-	guestInfo.VersionId = osInfo.VersionId
-	guestInfo.KernelVersion = osInfo.KernelVersion
-	guestInfo.Machine = osInfo.Machine
-	guestInfo.Id = osInfo.Id
-	return guestInfo
-}
+	// do the first round to fill the cache immediately
+	poll(p.AgentCommands)
 
-func (p *AgentPoller) UpdateDomain(domain *api.Domain) {
-	if domain != nil {
+	for {
 		select {
-		case p.domainUpdate <- domain:
-		default:
-			log.Log.Error("Failed to update agent poller domain info")
+		case <-closeChan:
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			poll(p.AgentCommands)
 		}
 	}
 }
 
-func (p *AgentPoller) createDomainData(domain *api.Domain) *DomainData {
-	aliasByMac := map[string]string{}
-	for _, ifc := range domain.Spec.Devices.Interfaces {
-		mac := ifc.MAC.MAC
-		alias := ifc.Alias.Name
-		aliasByMac[mac] = alias
-	}
-	return &DomainData{
-		name:       domain.Spec.Name,
-		aliasByMac: aliasByMac,
+type AgentPoller struct {
+	Connection cli.Connection
+	VmiUID     types.UID
+	domainName string
+	agentDone  chan struct{}
+	workers    []PollerWorker
+	agentStore *AgentStore
+}
+
+// CreatePoller creates the new structure that holds guest agent pollers
+func CreatePoller(
+	connecton cli.Connection,
+	vmiUID types.UID,
+	domainName string,
+	store *AgentStore,
+	qemuAgentPollerInterval time.Duration,
+) *AgentPoller {
+	p := &AgentPoller{
+		Connection: connecton,
+		VmiUID:     vmiUID,
+		domainName: domainName,
+		agentStore: store,
+		workers:    []PollerWorker{},
 	}
-}
-
-func (p *AgentPoller) pollQemuAgent(domainName string) (string, error) {
-	cmdResult, err := p.Connection.QemuAgentCommand("{\"execute\":\"guest-network-get-interfaces\"}", domainName)
-	return cmdResult, err
-}
 
-func (p *AgentPoller) pollQemuAgentForOsInfo(domainName string) (string, error) {
-	cmdResult, err := p.Connection.QemuAgentCommand("{\"execute\":\"guest-get-osinfo\"}", domainName)
-	return cmdResult, err
-}
+	// this have to be done via configuration passed in, for now this is OK
+	p.workers = append(p.workers, PollerWorker{
+		CallTick:      qemuAgentPollerInterval,
+		AgentCommands: []AgentCommand{GET_INTERFACES, GET_OSINFO},
```

Comment:
Would it make sense to have these as a parameter to `CreatePoller`?

I also assume that if there are ever multiple pollers, they will not be handling the same commands?

Reply:
Enabling the full configuration does not make sense. There will be only 4 polling workers. Each worker have its own special group (inherited from oVirt). Next is taken from PR: #2827 

* sys data: INTERFACES, OSINFO, TIMEZONE, HOSTNAME; queried every: 120s
* Agent: AGENT_VERSION; queried every: 300s
* Filesystem: FILESYSTEM; queried every: 300s
* User list: USERS: queried every: 10s

Enabling user to configure the choice of distribution does not make sense.
The calls are expensive and timeouts are chosen by trial in oVirt to not stress guest agent.

Also. These pollers are mutually exclusive and does not affect each other. This is the reason, why I selected the `async.Map`. Taken from the documentation:
```
The Map type is optimized for two common use cases: 
(1) when the entry for a given key is only ever written once 
but read many times, as in caches that only grow, or 
(2) when multiple goroutines read, write, and overwrite entries for
disjoint sets of keys. In these two cases, use of a Map may 
significantly reduce lock contention compared to 
a Go map paired with a separate Mutex or RWMutex. 
``` 
Case present in this PR and PR: #2827 is case (2).